crypto/fipsmodule/cipher/e_aesccm.c

```diff
@@ -112,8 +154,8 @@ static int ccm128_init_state(const struct ccm128_context *ctx,
     state->nonce.c[0] |= 0x40;  // Set AAD Flag
   }
   OPENSSL_memcpy(&state->nonce.c[1], nonce, nonce_len);
-  for (unsigned i = 0; i < L; i++) {
-    state->nonce.c[15 - i] = (uint8_t)(plaintext_len >> (8 * i));
+  for (uint32_t i = 0; i < L; i++) {
+    state->nonce.c[15 - i] = (uint8_t)((uint64_t) plaintext_len >> (8 * i));
```

Comment:
I spent a ridiculous amount of time figuring out that this explicit `uint64_t` cast is required. If `size_t` is 32-bits you end up shifting out of bounds like you said and that is undefined in the C standard. I observed that this somehow turns into a rotation on some platforms and you end up with the LSB shifted into the 5th byte position 🤷 . Since this routine is building up the 0th block of the MAC by encoding the message length into it, we need to properly encode a 0 for all the higher order bit positions or else the tag calculation gets hosed.

Here's an example I pulled from my notes on what the bad block looked like:

```
# After init state on a 32-bit machine
# This is wrong, we have 0x18 in the 5th byte position for some reason
(gdb) x/16b state
0xd40bfe4:  0x07  0x5a  0x8a  0xa4  0x85  0xc3  0x16  0xe9
0xd40bfec:  0x00  0x00  0x00  0x18  0x00  0x00  0x00  0x18

# Should be
(lldb) x/16b ccm_state
0x7ff24e871b08: 0x07 0x5a 0x8a 0xa4 0x85 0xc3 0x16 0xe9
0x7ff24e871b10: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x18
```